libzmix/src/signatures/bbs/signature.rs

```diff
@@ -0,0 +1,281 @@
+    pub fn get_unblinded_signature(&self, blinding: &FieldElement) -> Self {
+        Signature {
+            a: self.a.clone(),
+            s: self.s.clone() + blinding,
+            e: self.e.clone(),
+        }
+    }
+
+    // Verify a signature. During proof of knowledge also, this method is used after extending the verkey
+    pub fn verify(&self, messages: &[FieldElement], verkey: &PublicKey) -> Result<bool, BBSError> {
+        check_verkey_message(messages, verkey)?;
+        let b = compute_b_var_time(&G1::new(), verkey, messages, &self.s, 0);
+        let a = (&G2::generator() * &self.e) + &verkey.w;
+        Ok(GT::ate_2_pairing(&self.a, &a, &(-&b), &G2::generator()).is_one())
+    }
+}
+
+fn prep_vec_for_b(
+    public_key: &PublicKey,
+    messages: &[FieldElement],
+    blinding_factor: &FieldElement,
+    offset: usize,
+) -> (G1Vector, FieldElementVector) {
+    let mut points = G1Vector::with_capacity(messages.len() + 2);
+    let mut scalars = FieldElementVector::with_capacity(messages.len() + 2);
+    // XXX: g1 should not be a generator but a setup param
+    // prep for g1*h0^blinding_factor*hi^mi.....
+    points.push(G1::generator());
+    scalars.push(FieldElement::one());
+    points.push(public_key.h0.clone());
+    scalars.push(blinding_factor.clone());
+
+    for i in 0..messages.len() {
+        points.push(public_key.h[offset + i].clone());
+        scalars.push(messages[i].clone());
+    }
+    (points, scalars)
+}
+
+pub fn compute_b_const_time(
+    starting_value: &G1,
+    public_key: &PublicKey,
+    messages: &[FieldElement],
+    blinding_factor: &FieldElement,
+    offset: usize,
+) -> G1 {
+    let (points, scalars) = prep_vec_for_b(public_key, messages, blinding_factor, offset);
+    starting_value + points.multi_scalar_mul_const_time(&scalars).unwrap()
+}
+
+pub fn compute_b_var_time(
+    starting_value: &G1,
+    public_key: &PublicKey,
+    messages: &[FieldElement],
+    blinding_factor: &FieldElement,
+    offset: usize,
+) -> G1 {
+    let (points, scalars) = prep_vec_for_b(public_key, messages, blinding_factor, offset);
+    starting_value + points.multi_scalar_mul_var_time(&scalars).unwrap()
+}
+
+fn check_verkey_message(messages: &[FieldElement], verkey: &PublicKey) -> Result<(), BBSError> {
+    if messages.len() != verkey.message_count() {
+        return Err(BBSError::from_kind(
+            BBSErrorKind::SigningErrorMessageCountMismatch(verkey.message_count(), messages.len()),
+        ));
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::keys::generate;
+    use super::super::pok_sig::ProverCommittingG1;
+    use super::*;
+
+    #[test]
+    fn signature_serialization() {
+        let sig = Signature {
+            a: G1::random(),
+            e: FieldElement::random(),
+            s: FieldElement::random(),
+        };
+        let bytes = sig.to_bytes();
+        assert_eq!(bytes.len(), GroupG1_SIZE + MODBYTES * 2);
```

Comment:
Once again, a description of with this assertion matters would be helpful for those who haven't read the paper.